### DISCUSSION TITLE
In-browser calculator

### DISCUSSION THREAD

**Discussion by lubiedo** [minbrowser/min#1682] [Show and tell]
This is just an idea I had, inspired by some Android launchers. An inline calculator that will display the result along with search suggestions, places and instant answers. This way you just type the calculation and receive the result which you can copy by just clicking on it.

Demo:

https://user-images.githubusercontent.com/63729414/125837357-6e4641f0-94b2-462c-bd82-8f8a7cddd2fe.mov

It uses [mathjs](mathjs.org) to evaluate the expressions and avoid using any eval.

[Branch to try it out](https://github.com/lubiedo/min/tree/calc-plugin)

**Comment by lubiedo** [comment]
Another thing to mention is that mathjs also works with unit conversion :D
![image](https://user-images.githubusercontent.com/63729414/125984934-0dc6fabf-53f6-49be-a4c3-d5e7323d9232.png)

**Comment by shalva97** [comment]
it looks cool but there are other tools already. searching both "2 MiB to KiB" and (2+2)*123 to Google or DuckDuckGo shows the results.

**Reply by lubiedo** [reply]
Thanks for taking a look, @shalva97! Yes, definitely there are other resources but the idea behind this is that the user can avoid navigating to or using a website, or to make quick calculations without leaving a page or opening another tab.

**Reply by PalmerAL** [reply]
Yeah I agree with Lubiedo here, it is more convenient to have the answer right in the searchbar, rather than having to search for it. And calculations are frequently-used enough that there is a decently-large benefit to making them easier.

**Comment by PalmerAL** [comment]
With mathjs, I do worry a bit about security. They have a [security guide](https://github.com/josdejong/mathjs/blob/47a493d70ac92daa47b26f9eea88513a7efa4408/docs/expressions/security.md) which says it:

>  creates a fast performing function which can be used to evaluate the expression repeatedly using a dynamically passed scope.

...which means they're compiling math expressions into JS code. If you could trick the library into generating malicious JS somehow,  it could then be evaluated, and the UI process of the browser has node integration enabled, so it has full access to the user's computer.

I wonder if there's a simpler math parser we could use that would be safer?

**Reply by lubiedo** [reply]
Nice, thanks for pointing that out. I've followed their suggestions on that security guide and believe that it should be enough to avoid any kind of function utilization inside the parsing/evaluation. You can take a look here: https://github.com/lubiedo/min/commit/86a4e60647f32e0b359038c4454d44affdf4d6f9
Also, AFAIK since version >=4 they are not using `eval` in any kind or form.

**Reply by PalmerAL** [reply]
> they are not using eval in any kind or form.

that’s good!

Combined with the regex that limits this to only the characters in an expression, this seems reasonably safe.
